home/trainers.asm

```diff
@@ -229,13 +229,220 @@ FacingPlayerDistance::
 	call GetSpriteDirection
 	cp e
 	jr nz, .NotFacing
+
+	push de
+	ld hl, OBJECT_MAP_X ; x
+	add hl, bc
+	ld d, [hl]
+
+	ld hl, OBJECT_MAP_Y ; y
+	add hl, bc
+	ld e, [hl]
+
+	cp OW_LEFT
+	jr z, .check_boulder_left
+	cp OW_RIGHT
+	jr z, .check_boulder_right
+	cp OW_UP
+	jr z, .check_boulder_up
+	jmp .check_boulder_down
+
+.is_facing
 	scf
 	ret
 
 .NotFacing:
 	and a
 	ret
 
+.check_boulder_left
+	; trainer x = d, y = e
+	push bc
+	ld bc, wObject1Struct
+	ld hl, OBJECT_MAP_X
+	add hl, bc
+.check_boulder_left_loop
+	; boulder would need to be at the left of the trainer
+	ld a, [hl]
+	cp d
+	jr nc, .next_left_check
+	; boulder would need to be to the right of the player
+	ld hl, wPlayerMapX
+	cp [hl]
+	jr c, .next_left_check
+	; boulder would need to be at the same y as the trainer
+	ld hl, OBJECT_MAP_Y
+	add hl, bc
+	ld a, [hl]
+	cp e
+	jr nz, .next_left_check
+	; check for boulder sprite
+	ld hl, OBJECT_SPRITE
+	add hl, bc
+	ld a, [hl]
+	cp SPRITE_BOULDER_ROCK_FOSSIL
+	jr nz, .next_left_check
```

Comment:
I think *if* we do this at all, it should be a general-purpose "no trainer sight if there's any NPC in the way" check. (And even then, I'd prefer to redesign maps so the circumstance can't occur, if possible.)